core/Pos/Util/Util.hs

```diff
@@ -410,3 +429,103 @@ mkMinMax a = _MinMax # Just (a, a)
 
 minMaxOf :: Getting (MinMax a) s a -> s -> Maybe (a, a)
 minMaxOf l = view _MinMax . foldMapOf l mkMinMax
+
+----------------------------------------------------------------------------
+-- Filesystem utilities
+----------------------------------------------------------------------------
+
+-- | Lists all immediate children of the given directory, excluding "." and ".."
+-- Returns all the files inclusive of the initial `FilePath`.
+ls :: MonadIO m => FilePath -> m [FilePath]
+ls initialFp = map ((</>) initialFp) <$> liftIO (listDirectory (normalise initialFp))
+
+-- | Lists all recursive descendants of the given directory.
+lstree :: MonadIO m => FilePath -> m [FilePath]
+lstree fp = go mempty fp
+  where
+    consUniq :: FilePath -> [FilePath] -> [FilePath]
+    consUniq x xs = if x /= fp then (x : xs) else xs
+
+    go :: MonadIO m => [FilePath] -> FilePath -> m [FilePath]
+    go !acc currentFp = do
+        isDirectory <- liftIO (doesDirectoryExist currentFp)
+        case isDirectory of
+            True  -> ls currentFp >>= foldM go (consUniq currentFp acc)
+            False -> return (consUniq currentFp acc)
+
+-- | Creates a temporary directory, nuking it after the inner action completes,
+-- even if an exception is raised.
+withTempDir :: FilePath
+            -- ^ Parent directory
+            -> Text
+            -- ^ Directory name template
+            -> (FilePath -> IO a)
+            -> IO a
+withTempDir parentDir template = bracket acquire dispose
+  where
+    acquire :: IO FilePath
+    acquire = do
+        tid <- myThreadId
+        pth <- canonicalizePath $ normalise $ parentDir </> (toString template <> show tid)
+        createDirectory pth
+        return pth
+
+    dispose :: FilePath -> IO ()
+    dispose = removeDirectory
+
+-- | Simple shim to emulate the behaviour of `Filesystem.Path.directory`,
+-- which is a bit more lenient than `System.FilePath.takeDirectory`.
+directory :: FilePath -> FilePath
+directory "" = ""
+directory f = case last f of
+    '/' -> f
```

Comment:
You are absolutely right! In fact, I think we should be using [pathSeparator](http://hackage.haskell.org/package/filepath-1.4.1.2/docs/src/System.FilePath.Posix.html#pathSeparator) here (sorry for the uninformative link, but apparently Hackage is having some trouble showing us the real Haddock due to CPP), but that has essentially signature `pathSeparator :: Char` which is OS-specific and that we can use to infer the same thing in a platform-independent way.